lib/knuckle_cluster.rb

```diff
@@ -185,12 +185,20 @@ def generate_connection_string(agent:, subcommand: nil, port_forward: nil)
     end
 
     def agent_registry
-      @agent_registry ||= AgentRegistry.new(
-        aws_client_config: aws_client_config,
-        cluster_name:      cluster_name,
-      )
+      @agent_registry ||= ()
+      if @cluster_name
+        EcsAgentRegistry.new(
+          aws_client_config: aws_client_config,
+          cluster_name:      cluster_name,
+        )
+      elsif @spot_request_id
+        SpotRequestInstanceRegistry.new(
+          aws_client_config: aws_client_config,
+          spot_request_id:   spot_request_id,
+        )
+      end
```

Comment:
Don't think this will memoize properly with the `()` at the top, so you'll be recreating the registry with every call. Suggest breaking up the registry generation functions like this for readability:

```
    def agent_registry
      @agent_registry ||= if cluster_name
        ecs_agent_registry
      elsif spot_request_id
        spot_request_instance_registry
      end
    end

    def ecs_agent_registry
      EcsAgentRegistry.new(
        aws_client_config: aws_client_config,
        cluster_name:      cluster_name,
      )
    end

    def spot_request_instance_registry
      SpotRequestInstanceRegistry.new(
        aws_client_config: aws_client_config,
        spot_request_id:   spot_request_id,
      )
    end
```

(or just surrounding the whole thing in a dirty big `begin...end` block)

Reply:
Good pickup - this was a typo.